src/types.ts

```diff
@@ -1,3 +1,36 @@
+import { render } from "@testing-library/react";
+import { mount } from "enzyme";
+
+export type FullProps<C extends React.ComponentType> = React.ComponentProps<C>;
+
+type RequiredKeys<T> = {
+  [K in keyof T]-?: Record<string, unknown> extends { [P in K]: T[K] } ? never : K;
+}[keyof T];
+
+export type HasRequiredField<T> = RequiredKeys<T> extends never ? false : true;
+
+export type RenderEnzyme<C extends React.ComponentType, P extends Partial<FullProps<C>>> = (
+  ...testProps: HasRequiredField<RemainingPropsAndTestOverrides<C, P>> extends true
+    ? [RemainingPropsAndTestOverrides<C, P>]
+    : [RemainingPropsAndTestOverrides<C, P>?]
+) => RenderEnzymeReturn<C>;
```

Comment:
This, too, is the other part of the magic, changing the return type based on the type of the given default props.